thunder/executors/fa3ex.py

```diff
@@ -0,0 +1,156 @@
+import torch
+
+# flash attn 3
+try:
+    from flash_attn_interface import _flash_attn_forward, _flash_attn_backward, flash_attn_func
+
+    HAS_FA3 = True
+except:
+    HAS_FA3 = False
+
+import thunder
+from thunder.core.transforms import get_grad, put_grads
+from thunder.extend import OperatorExecutor, register_executor
+
+fa3_ex: OperatorExecutor = OperatorExecutor("fa3", version="0.1")
+register_executor(fa3_ex)
+
+
+def fa3_fwd_meta(
+    q: thunder.torch.TensorLike,
+    k: thunder.torch.TensorLike,
+    v: thunder.torch.TensorLike,
+    causal: bool = False,
+    softmax_scale: float | None = None,
+):
+    return thunder.TensorProxy(like=q), thunder.TensorProxy(like=q, shape=(*q.shape[:1], 1))
+
+
+def fa3_fwd_impl(
+    q: torch.Tensor, k: torch.Tensor, v: torch.Tensor, causal: bool = False, softmax_scale: float | None = None
+):
+    if not HAS_FA3:
+        raise Exception("fa3 not built, cannot use fa3 executor")  # checker should fail before getting here
+
+    if softmax_scale is None:
+        softmax_scale = q.shape[-1] ** (-0.5)
+    out, q, k, v, out_padded, softmax_lse, S_dmask = _flash_attn_forward(q, k, v, softmax_scale, causal)
```

Comment:
You mentioned elsewhere that the inputs to the operation must be contiguous. Would you add a comment about that requirement, and then call `torch.contiguous` on the inputs that must be contiguous here?

Would you also add a test with noncontiguous inputs, and verify the executor works as expected?

fyi @IvanYashchuk @jjsjann123 -- particular stride requirements are likely to come up more in the future. Instead of putting the contiguous calls into the implementation, it'd be nice if we could put them into the transformation for the function. But if we do that today then it will always trigger a copy, even if the input tensors are always contiguous

Reply:
> particular stride requirements are likely to come up more in the future. Instead of putting the contiguous calls into the implementation, it'd be nice if we could put them into the transformation for the function.

Does having it in the transformation mean we'll have a `stride_order` node in the trace?
Admittedly, having a `contiguous` call could be tricky, since it can be executed as a view or a copy. The ambiguity is tricky to handle if we want to have consistent behavior with eager.

> But if we do that today then it will always trigger a copy, even if the input tensors are always contiguous

If torch executor takes the contiguous call and figured it's a no-op, this will only produce a view to the buffer. So I think this statement is on nvfuser's handling of stride_order? But if that's fused, we aren't really making any real copy...

Reply:
I think that if it's in the transformation then yes, `stride_order` will be in the trace. 

Yes you're also right that we could check that our `stride_order` execution can no-op, but currently it doesn't:

https://github.com/Lightning-AI/lightning-thunder/blob/67ec2c6f4e72f6215f12941bf161915f8f458181/thunder/executors/torchex.py#L621

Reply:
Awesome, thanks for working that through with me!

So, @Fuzzkatt, what about putting `contiguous` calls in the transform, to ensure that the inputs are the expected format, and adding a test that noncontiguous inputs work as expected?

Reply:
Sure I can add the contiguous calls in the transform and add a test. One caveat: are we concerned about the additional overhead for H2D memcpy that a .contiguous() convert might add that was brought up in https://github.com/Dao-AILab/flash-attention/issues/1087? I can also try to run some timing tests if it would help

Reply:
You're correct that would be unfortunate, and in the future we'll want to do a better job selecting which executor to use (possibly with some benchmarking tools), but this PR doesn't need to worry about that. Practitioners can always disable the fa3 executor globally as a performance workaround. 

This is a good issue to leave for a follow-up.

Reply:
Added the contiguous() calls in both the fwd and bwd, as they ended up also being necessary in the bwd. The weird thing is that the actual _flash_attn_forward() and _flash_attn_backward() calls from fa3 itself that I'm calling into in the fwd_impl() and bwd_impl() of the thunder fa3 executor should already be doing these contiguous() calls: https://github.com/Dao-AILab/flash-attention/blob/main/hopper/flash_attn_interface.py#L18 and https://github.com/Dao-AILab/flash-attention/blob/main/hopper/flash_attn_interface.py#L44. But without adding these explicit calls in the transforms, I'm still hitting the 'Inputs must be contiguous' RuntimeError. I'll keep looking into it tomorrow.

Reply:
That is curious! 

I think we find that determining exactly when a kernel (or set of kernels) can be used is often the hardest part of integrating them with Thunder. fyi @tfogal and @IvanYashchuk. 

There are two thoughts here I think the project needs to be aware of:

First, incorporating this logic about when a particular kernel can and should be used is part of the Thunder project, and it's good we're very thorough, and in the future we'll have benchmarking facilities to help us make decisions about which executor to use.

Second, even though Thunder is intended to select the best kernels given the current program, we probably need to think more about letting practitioners know where they are losing performance. We have seen this issue in the past, where program was calling flash attention in a way that didn't use our fastest kernels, and this caused a huge performance impact and a lot of end user confusion.

I expect statistics on how operations are called, analysis of performance gaps between executors, automated benchmarking, and descriptions of how practitioners could utilize faster kernels will be part of thunder's Q4 planning.